components/canvas/webgl_thread.rs

```diff
@@ -1091,147 +1137,138 @@ impl WebGLImpl {
             },
             WebGLCommand::GetShaderInfoLog(shader_id, ref chan) => {
-                Self::shader_info_log(ctx.gl(), shader_id, chan)
+                Self::shader_info_log(gl, shader_id, chan)
             },
             WebGLCommand::GetProgramInfoLog(program_id, ref chan) => {
-                Self::program_info_log(ctx.gl(), program_id, chan)
+                Self::program_info_log(gl, program_id, chan)
             },
             WebGLCommand::CompileShader(shader_id, ref source) => {
-                Self::compile_shader(ctx.gl(), shader_id, &source)
+                Self::compile_shader(gl, shader_id, &source)
             },
-            WebGLCommand::CreateBuffer(ref chan) => Self::create_buffer(ctx.gl(), chan),
-            WebGLCommand::CreateFramebuffer(ref chan) => Self::create_framebuffer(ctx.gl(), chan),
-            WebGLCommand::CreateRenderbuffer(ref chan) => Self::create_renderbuffer(ctx.gl(), chan),
-            WebGLCommand::CreateTexture(ref chan) => Self::create_texture(ctx.gl(), chan),
-            WebGLCommand::CreateProgram(ref chan) => Self::create_program(ctx.gl(), chan),
+            WebGLCommand::CreateBuffer(ref chan) => Self::create_buffer(gl, chan),
+            WebGLCommand::CreateFramebuffer(ref chan) => Self::create_framebuffer(gl, chan),
+            WebGLCommand::CreateRenderbuffer(ref chan) => Self::create_renderbuffer(gl, chan),
+            WebGLCommand::CreateTexture(ref chan) => Self::create_texture(gl, chan),
+            WebGLCommand::CreateProgram(ref chan) => Self::create_program(gl, chan),
             WebGLCommand::CreateShader(shader_type, ref chan) => {
-                Self::create_shader(ctx.gl(), shader_type, chan)
-            },
-            WebGLCommand::DeleteBuffer(id) => ctx.gl().delete_buffers(&[id.get()]),
-            WebGLCommand::DeleteFramebuffer(id) => ctx.gl().delete_framebuffers(&[id.get()]),
-            WebGLCommand::DeleteRenderbuffer(id) => ctx.gl().delete_renderbuffers(&[id.get()]),
-            WebGLCommand::DeleteTexture(id) => ctx.gl().delete_textures(&[id.get()]),
-            WebGLCommand::DeleteProgram(id) => ctx.gl().delete_program(id.get()),
-            WebGLCommand::DeleteShader(id) => ctx.gl().delete_shader(id.get()),
-            WebGLCommand::BindBuffer(target, id) => ctx
-                .gl()
-                .bind_buffer(target, id.map_or(0, WebGLBufferId::get)),
+                Self::create_shader(gl, shader_type, chan)
+            },
+            WebGLCommand::DeleteBuffer(id) => gl.delete_buffers(&[id.get()]),
+            WebGLCommand::DeleteFramebuffer(WebGLFramebufferId::Transparent(id)) => {
+                gl.delete_framebuffers(&[id.get()])
+            },
+            WebGLCommand::DeleteFramebuffer(WebGLFramebufferId::Opaque(
+                WebGLOpaqueFramebufferId::WebXR(_),
+            )) => {
+                // TODO: deleting a WebXR framebuffer (which happens when the framebuffer is GCd)
```

Comment:
OK, the issue here is what happens if the XRSession gets GC'd before it's ended. The XR session is still running, so I was concerned that we'd end up removing the swap chain from the XR device when it wasn't expecting it. On reflection, we're OK since the swap chain is an `Arc` under the hood, so won't get dropped until both the webgl thread and the xr thread are done with it. So we can just delete the entry from the hash table.